[Workflow] Mark registry as internal and deprecate the service

lyrixx opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | yes
| Tickets       | Fix #45315
| License       | MIT
| Doc PR        | 

---

The Registry was added to be able to get a workflow easily from twig templates
without having to specify the workflow name.

Unfortunately, many people thought this was the only way to get a workflow and
we even add some flexibility to retrieve the workflow from the registry.

I think it's a bad idea to inject the registry, it has the same default as
injecting the Container: It's a black box, does not respect SOLID principle and
the demeter law, and make testing harder.

So I decided to mark the registry as internal. Instead people have to inject the
"right" workflow where they need it.

It can be legit to need all workflows, for documentation or for testing.
So, all workflow services are tagged and can be injected with the
following YAML syntax:

```yaml
!tagged_locator  { tag: workflow, index_by: name }
```

or PHP syntax:

```php
tagged_locator('workflow', 'name')
```

Also, two others tags exists for each workflow types
* `workflow.workflow`
* `workflow.state_machine`

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1017,8 +1016,9 @@ private function registerWorkflowConfiguration(array $config, ContainerBuilder $
-        $commandDumpDefinition->setArgument(0, $workflows);
+        $workflowLocator = (string) ServiceLocatorTagPass::register($container, $workflowLocator);
+        $container->setAlias('workflows', $workflowLocator);
+        $container->registerAliasForArgument($workflowLocator, ServiceLocator::class, 'workflows');
```
what about `Psr\Container\ContainerInterface instead`? is there any reason to force `ServiceLocator`?

but more importantly: what about removing this alias altogether and let ppl use `tagged_locator('workflow')` after tagging all workflows with ... `workflow`?

lyrixx replied:
I chose a `ServiceLocator` because it is what it really is.

More over, on the `ContainerInterface`, we miss the method to get all Workflow (needed in the Command). And people might need it too.

---

> tagged_locator('workflow')

I **really love** this feature, but it it not so well know. I fear people will miss this part. The discoverability is a bit worst: You need to know all workflows are tagged with `workflow` + how `tagger_locator` works. But it's kind of better indeed, since we don't need to add a new hardcoded id in the container. and we could even have 3 tags `workflows` (all) `workflow.workflow` (just type=workflow) `workflows.state_machine` (type=state_machine)


What do you think?

nicolas-grekas replied:
Good idea, let's do it. I very much prefer this over a named autowiring alias in this case. Let's document this properly and ppl will discover it.

lyrixx replied:
Done 👍🏼

nicolas-grekas reviewed (commented):
Makes sense to me.
we should also deprecate the corresponding autowiring alias:
src/Symfony/Bundle/FrameworkBundle/Resources/config/workflow.php:        ->alias(Registry::class, 'workflow.registry')

94noni commented on src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php:
```diff
@@ -40,19 +42,28 @@ class WorkflowDumpCommand extends Command
+        if (is_array($workflows)) {
+            $this->definitions = $workflows;
+            trigger_deprecation('symfony/framework-bundle', '6.2', 'Passing an array of definitions in "%s()" is deprecated. Inject a ServiceLocator filled with all workflows instead.', __METHOD__);
+        } elseif ($workflows instanceof ServiceLocator) {
```
why not this as first if, as it will be the "new code"

lyrixx replied:
Fixed

fabpot commented:
@lyrixx Do you have time to finish this one?

lyrixx commented:
Yes, I'll do that ASAP, But I'm on holidays for 2 weeks. I'll do that after!

lyrixx committed:
[Workflow] Mark registry as internal and deprecate the service

Instead, all workflow services are tagged and can be injected with the
following YAML syntax:

```yaml
!tagged_locator  { tag: workflow, index_by: name }
```

or PHP syntax:

```php
tagged_locator('workflow', 'name')
```

Also, two others tags exists for each workflow types
* `workflow.workflow`
* `workflow.state_machine`

lyrixx commented:
Hello,

I pushed a new version without the service locator. I also updated the PR description with the new usage

fabpot commented:
Thank you @lyrixx.

tacman commented:
I'm in the category of programmers that injected the workflow registry in order to get a list of workflows in a helper bundle for workflow.

```php
// now deprecated, following https://symfony.com/doc/current/workflow.html#accessing-the-workflow-in-a-class

class WorkflowHelperService
{
    public function __construct(
        private Registry $workflowRegistry,

// in loadExtension of the bundle:
        $builder->autowire( WorkflowHelperService::class)
            ->setArgument('$workflowRegistry', new Reference('workflow.registry'));
       
      
```

```php

// better way
class WorkflowHelperService
{
    public function __construct(
        private ServiceLocator $locator,

// loadExtension()

        $builder->autowire( WorkflowHelperService::class)
            ->setArgument('$locator', tagged_locator(tag: 'workflow', indexAttribute: 'name' ))
            ->setAutoconfigured(true)
        ;

```

Using this, I'm not getting any services in the locator, $this->locator->count() is zero.

I imagine I'm not using tagger_locator correctly, or I'm doing this in the wrong spot (I'm doing this in loadExtension of my bundle, perhaps it needs to be done in the compiler pass?  If so, could you point me in the right direction on how to do that, I still find the bundle interactions a bit intimidating.

tacman commented:
Also, is this the right spot for that question?  The issue itself is closed, should I open a new one?

stof commented:
You should open a new discussion
